Added utility classes, reworked BindingAdapters, updated ViewModel test

Aidanvii7 opened the pull request:
Added utility classes from my project to avoid  the dependency.
Reworked binding adapters to still use rxBindings internally, but the is hidden from the view model.

Aidanvii7 commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/users/UserEditViewModel.kt:
```diff
@@ -1,61 +1,89 @@
-        compositeDisposable.add(obs.subscribe(firstNameRelay))
+    //endregion
+
+    //region bindable properties ( view << view model )
```
I added these purely for learning purposes. 
With MVVM, understanding the flow of of information between the view and view model is important. 
I can remove if you think it's overkill.

exallium replied:
I don't think it's overkill but lets do our best to follow consistent patterns:

* companion object
* public fields
* private fields
* public methods
* private methods

as closely as possible

Aidanvii7 replied:
agreed, it's hard to see where regular Kotlin properties fit into this model though.

Aidanvii7 commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/users/UserEditViewModel.kt:
```diff
@@ -1,61 +1,89 @@
 
-    val firstNameError = ObservableField<String>("")
-    val lastNameError = ObservableField<String>("")
+    private var disposable by ecoDisposable
```
notifies Disposable delegate on assignment, so previous value is disposed of

Aidanvii7 commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/users/UserEditViewModel.kt:
```diff
@@ -1,61 +1,89 @@
+    private var disposable by ecoDisposable
+    private var savedUser by savedUserRelay
+    private var firstName by firstNameRelay
+    private var lastName by lastNameRelay
```
notifies relay delegates on assignment, which forwards to actual relay

exallium commented on app/src/main/java/com/exallium/mvvmcleanexample/app/binding/TextInputEditTextBindings.kt:
```diff
@@ -3,16 +3,20 @@ package com.exallium.mvvmcleanexample.app.binding
-        consumer.accept(textInputEditText.textChanges().map(CharSequence::toString))
+    @BindingAdapter("onTextChanged")
+    fun onTextChanged(textInputEditText: TextInputEditText, consumer: StringConsumer?) {
+        (textInputEditText.tag as? Disposable)?.dispose()
```
I think this is a good change.

If the view model out-lives the view, say, due to a configuration change or something, we don't end up leaking the view.  I think.  If this is being done for such a change, that's cool.  I'm trying to think of ways which utilising tag might be unsafe.

Aidanvii7 replied:
I thought so too about the tag, checking in the data binding framework code they seem to use it. There's probably other ways to do this, but this was the fastest.

exallium replied:
If there's a precedent of utilizing the tag for something like this in the binding framework, then I guess that's ok.  I'm curious though about applying multiple bindings to a single view, and whether you'd get clobbering.

exallium commented on app/src/main/java/com/exallium/mvvmcleanexample/app/users/UserEditView.kt:
```diff
@@ -25,34 +26,46 @@ class UserEditView : AppCompatActivity() {
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
-        val binding = DataBindingUtil.setContentView<UserEditViewBinding>(this, R.layout.user_edit_view)
+        binding = DataBindingUtil.setContentView<UserEditViewBinding>(this, R.layout.user_edit_view).also { binding ->
```
what's with the also block?

Aidanvii7 replied:
binding is mutable, so smart cast is impossible.

exallium replied:
isn't the binding strongly typed via `setContentView<UserEditViewBinding>` ?

Aidanvii7 replied:
it is, but it is defined as a nullable var: 

```
private var binding by autoEcoBinding<UserEditViewBinding>(null)
// without inference
private var binding: UserEditViewBinding? by autoEcoBinding<UserEditViewBinding>(null)
```
The reason for the also block means that it is safely cast to non null type.
The binding wont be written to by another thread of course, I just wish I could tell kotlin that so it just smart casts anyway :/

exallium replied:
is there any way we can set the binding then do the rest of setup outside? I'm just not a huge fan of extra levels of scoping.

Aidanvii7 replied:
yea there's a way, I don't like it either.

exallium commented on app/src/main/java/com/exallium/mvvmcleanexample/app/users/UserEditView.kt:
```diff
@@ -25,34 +26,46 @@ class UserEditView : AppCompatActivity() {
-                is AppRouter.Transition.DisplayError -> displayError(binding.root, it.message)
+            router.transitions().subscribe {
+                when (it) {
+                    is AppRouter.Transition.GoBack -> {/*FIXME this causes a crash: onBackPressed()*/}
```
FIXME

Aidanvii7 replied:
when I run this it just crashes, is it supposed to close the key board or end the activity? I wasn't sure.

exallium replied:
It's supposed to "go back" to the previous screen.  in this case, that'd be onBackPressed.

exallium commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/users/UserEditViewModel.kt:
```diff
@@ -1,61 +1,89 @@
+import com.exallium.mvvmcleanexample.presentation.utils.ecoDisposable
+import com.exallium.mvvmcleanexample.presentation.utils.notifiable.BaseNotifiableObservable
+import com.exallium.mvvmcleanexample.presentation.utils.notifiable.NotifiableObservable
+import com.exallium.mvvmcleanexample.presentation.utils.publishRelay
```
we're bringing in a lot more than we are removing.  Makes me wonder whether we might be over-doing it.

Aidanvii7 replied:
Sure thing, though with this we can check that the relays are being notified from the tests, as we can pass in a mocked property delegate. Thoughts?

exallium replied:
we always have the option to observe something further down the chain.  My only concern here in a sample app is we might be hiding some of the mechanisms from the reader when they come into this file.  Again, I think these are very cool, but I wonder how deep down the rabbit hole we want to go for a sample app

exallium commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/utils/Extensions.kt:
```diff
@@ -0,0 +1,72 @@
+
+fun ecoDisposable(disposable: Disposable?): EcoDisposableProperty = EcoDisposableProperty(disposable)
+
+fun <T> publishRelay(initialValue: T? = null): RelayProperty<T> {
```
let's document all of these w. /** */ since we're starting to get into more advanced kotlin features (those that don't exist in Java) so that people not so familiar with kotlin can still follow along.

exallium commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/utils/PropertyCacher.kt:
```diff
@@ -0,0 +1,37 @@
+
+import kotlin.reflect.KProperty
+
+object PropertyCacher {
```
Javadoc.

What does this do for us? Why is it necessary?

To someone coming into the codebase, this reads like a bunch of boilerplate.

Aidanvii7 replied:
It is boilerplate, but it avoids having to reference BR.propertyName for each changed property in an Observable object (Observable from the data binding framework I mean).

exallium replied:
Ok cool lets add some docs to this class then

Aidanvii7 replied:
docs incoming

exallium commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/utils/delegates/EcoDisposableProperty.kt:
```diff
@@ -0,0 +1,15 @@
+import kotlin.properties.ReadWriteProperty
+import kotlin.reflect.KProperty
+
+class EcoDisposableProperty(private var disposable: Disposable?) : ReadWriteProperty<Any?, Disposable?> {
```
can you make this be composite instead of a one-off? I think disposable is a general enough concept that I'd like to make ViewModel disposable.

Aidanvii7 replied:
Can maybe do something with operator overloading or infix operators, but that's new territory for me. Will try.

exallium replied:
I'm even thinking of like a CompositeDisposableDelegate that you just add things to, and when you set it to null it clears it.  Not really sure what that'd look like.  in fact, it'd probably be easier to just use a CompositeDisposable.

exallium commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/utils/notifiable/BaseNotifiableObservable.kt:
```diff
@@ -0,0 +1,67 @@
+     */
+    @Synchronized
+    override fun notifyChange() {
+        if (initialized) {
```
why do we need this separate flag?

Aidanvii7 replied:
I can remove this class, we probably don't need it. This is merely a work around to not enforce having to inherit from the framework version of BaseObservable and still get all the features from it. This is a re-implementation of it that allows us to delegate the behaviour.

Actually the flag is not needed as long as I hold a reference to the Lazy property delegate (I think)

exallium replied:
the Lazy delegate should only ever be fired once, yes.  Generally, boolean flags like this feel like something funny is going on.

exallium commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/utils/notifiable/BaseNotifiableObservable.kt:
```diff
@@ -0,0 +1,67 @@
+import android.databinding.Observable
+import android.databinding.PropertyChangeRegistry
+
+class BaseNotifiableObservable : NotifiableObservable {
```
Javadoc to answer why we want this class

Aidanvii7 replied:
I can remove this class, we probably don't need it. This is merely a work around to not enforce having to inherit from the framework version of BaseObservable and still get all the features from it. This is a re-implementation of it that allows us to delegate the behaviour to it. That or you implement Observable yourself and reimplement the PropertyChangeRegistry boilerplate every time.

exallium replied:
I see.  Hmm.  I think if it's preventing repeated code, it's probably ok to keep around.  I have an adversion to boilerplate.

exallium commented on presentation/src/test/java/com/exallium/mvvmcleanexample/presentation/users/UserEditViewModelTest.kt:
```diff
@@ -19,32 +20,40 @@ import org.mockito.MockitoAnnotations
-    val saveClicks = PublishSubject.create<Unit>()
-    val firstNameChanges = PublishSubject.create<CharSequence>()
-    val lastNameChanges = PublishSubject.create<CharSequence>()
+    @Mock lateinit var model: UserEditModel
```
let's keep consistent:

```
@Mock
lateinit var model: UserEditModel
```

Aidanvii7 replied:
sure thing

exallium commented on presentation/src/test/java/com/exallium/mvvmcleanexample/presentation/users/UserEditViewModelTest.kt:
```diff
@@ -54,33 +63,122 @@ class UserEditViewModelTest {
+    }
+
+    @Test
+    fun `notifies savedUserRelay when SaveClicked`() {
```
these test names are nice..but do they work for android projects now? I didn't know they were working.

Regardless, we should be consistent, so we should choose one method and stick with it for all test names.

Aidanvii7 replied:
sure I will remove, based it on a talk that Danny did. They work in tests, you can disable the warning for tests in the IDE.

exallium replied:
Oh neat.  That's actually awesome, and makes for much more readable tests.  I'm fine with converting them all to this (keeping some kind of consistent naming format like) and adding a line in the readme about these warnings and how to turn them off, as i think it is useful information in general.

Aidanvii7 replied:
yep I agree, would be better to have them consistent, also good idea about a README.

exallium reviewed (changes requested):
Looks decent.  I'm wary of bringing in too much framework-ish code but if you can back up your proposed classes with solid grounds then I'm cool with it.

Let's make sure we don't lose focus of the goal of the repo, to be a clear, well defined, MVVM exmaple app based on Reactive programming and clean architecture.

Aidanvii7 commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/users/UserEditViewModel.kt:
```diff
@@ -1,61 +1,86 @@
+    // view << >> view model
+    @get:Bindable
+    var firstName: String by bindable(firstName) {
+        firstNameRelay.accept(it)
```
TODO replace with method reference

Aidanvii7 commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/users/UserEditViewModel.kt:
```diff
@@ -1,61 +1,86 @@
-                .subscribe(this::handleResult, this::handleError))
+    @get:Bindable
+    var lastName: String by bindable(lastName) {
+        lastNameRelay.accept(it)
```
TODO replace with method reference

exallium replied:
TODO

Aidanvii7 commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/users/UserEditViewModel.kt:
```diff
@@ -1,61 +1,86 @@
+
+    private var disposable: Disposable? by ecoDisposableProperty
+    private var savedUser: User by distinctObservable(User(firstName, lastName)) {
+        savedUserRelay.accept(it)
```
TODO replace with method reference

exallium replied:
TODO

Aidanvii7 commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/utils/Extensions.kt:
````diff
@@ -0,0 +1,159 @@
+ * @get:Bindable
+ * var firstName by bindable("")
+ * ```
+ * TODO replace with provideDelegate alternative to get the
````
whoops, this one slipped through!

Aidanvii7 commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/utils/notifiable/BaseNotifiableObservable.kt:
```diff
@@ -0,0 +1,75 @@
+
+     * @param propertyId The generated BR id for the [Bindable] field.
+     */
+    @Synchronized
```
TODO remove documentation

exallium commented on app/src/main/res/layout/user_edit_view.xml:
```diff
@@ -17,31 +19,31 @@
                 android:layout_height="wrap_content"
                 android:hint="@string/first_name"
-                app:bindTextChanges="@{viewModel::setFirstNameChanged}" />
+                bind:text="@={viewModel.firstName}"/>
```
Change of heart w.r.t. inverse bindings?

aidan-mcwilliams replied:
My problem with Inverse bindings is when both the view and view model can change the property, however if only 1 side is changing the property during the binding, it works. 

Will explain in the view model.

exallium commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/users/UserEditViewModel.kt:
```diff
@@ -1,61 +1,86 @@
+        private val firstNameRelay: BehaviorRelay<String> = BehaviorRelay.create(),
+        private val lastNameRelay: BehaviorRelay<String> = BehaviorRelay.create(),
+        private val savedUserRelay: PublishRelay<User> = PublishRelay.create(),
+        firstName: String = "",
```
I'm thinking we should actually be passing in a userId here, and then querying the UserRepository for the User object via the UserEditModel.  Reason being is single source of truth.

exallium commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/users/UserEditViewModel.kt:
```diff
@@ -1,61 +1,86 @@
-                .subscribe(clickRelay))
+    // view << view model
+    @get:Bindable
+    var firstNameError: String by bindable("")
```
would still like these to be at the top of the file.

exallium commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/utils/Extensions.kt:
```diff
@@ -0,0 +1,159 @@
+        : ReadWriteProperty<Any?, T> {
+
+    return if (bindableResourceId == USE_PROPERTY_CACHER) {
+        object : DistinctObservableProperty<T>(initialValue) {
```
code duplication -> extract into a subclass.

exallium commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/utils/bindingadapters/AdapterUtils.kt:
```diff
@@ -0,0 +1,25 @@
+typealias ListenerAction<V, Listener> = V.(Listener) -> Unit
+
+inline fun <V : View, L> V.trackListener(
+
```
space

exallium commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/utils/bindingadapters/TextViewBindingAdapters.kt:
```diff
@@ -0,0 +1,73 @@
+import android.widget.TextView
+import com.android.databinding.library.baseAdapters.R
+
+const val attrText = "bind:text"
```
Let's keep static final variable names in all caps.

aidan-mcwilliams replied:
alright will change

exallium commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/utils/bindingadapters/TextViewBindingAdapters.kt:
```diff
@@ -0,0 +1,73 @@
+    fun invoke(text: CharSequence)
+}
+
+@InverseBindingAdapter(
```
curious as to why we moved back to InverseBindingAdapter.

exallium commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/utils/bindingadapters/ViewBindingAdapters.kt:
```diff
@@ -0,0 +1,19 @@
+
+@BindingAdapter(value = attrOnClick)
+fun setOnClick(view: View, onClick: OnClick?) {
+    if (onClick != null) {
```
```view.setOnClickListener(if (onClick != null) {{ onClick.invoke() }} else null)```

aidan-mcwilliams replied:
syntax is invalid

exallium reviewed (changes requested):
Looks like there are some outstanding TODO's left here.  Also, I don't really like the idea of handing the ViewModel a pair of strings for first and last name, I think these are better off as a single userId that the ViewModel can query a User object for.

aidan-mcwilliams commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/users/UserEditViewModel.kt:
```diff
@@ -1,61 +1,93 @@
+    var firstName: String by distinctObservable(
+            initialValue = "",
+            onChange = firstNameRelay::accept)
 
```
This is used in the inverse binding. When the `UserEditViewModel` is passed to the `UserEditViewBinding`, the binding reads the `firstName` property once to get it's initial value. The reason it reads it once is because it's not annotated with `Bindable` - when you do this, the generated `ViewDataBinding` will listen for changes on that property, instead of reading it once on initialisation.

aidan-mcwilliams commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/users/UserEditViewModel.kt:
```diff
@@ -1,61 +1,93 @@
+    // view << view model
+    @get:Bindable
+    var firstNameError: String by bindable("")
+
```
Here the view model changes the `firstNameError` property during the lifetime of the binding, so the view needs notified of changes, so we annotate it with `Bindable`. It's the same effect as having an `ObservableField<String>` for each property, but now we have a single `PropertyChangeRegistry` under the hood, handled by the `BaseNotifiableObservable`.

exallium replied:
What benefits do this single registry give us? What problem are we solving?

Aidanvii7 replied:
Good question. Quite honestly, it's unclear. There are different ways of using data binding, implementing `Observable` is considered the most flexible, Extending the framework `BaseObservable` is a convenience as there is some boilerplate involved. 

See this section: https://developer.android.com/topic/libraries/data-binding/index.html#observable_objects
This too covers the approaches: https://medium.com/google-developers/android-data-binding-observability-9de4ff3fe038

What I don't like about `ObservableField` and the primitive counterparts is that they expose public setters on the properties of the view model that are only intended to be observed by the view, where the state is managed internally by the view model.

This is something I forgot to address in this PR but can be easily changed with:
```
@get:Bindable
var firstNameError: String by bindable("")
   private set 
```

exallium replied:
```
What I don't like about ObservableField and the primitive counterparts is that they expose public setters on the properties of the view model that are only intended to be observed by the view, where the state is managed internally by the view model.
```

If the View (and I guess the activity) are the only things getting access to the ViewModel, I'm not sure how relevant it is that they're public.

exallium commented on app/src/main/AndroidManifest.xml:
```diff
@@ -8,7 +8,7 @@
         android:supportsRtl="true"
         android:theme="@style/AppTheme">
-        <activity android:name=".app.users.UserEditView">
+        <activity android:name=".app.users.UserEditActivity">
```
Now that I think about it, this is a much better name.  Because the UserEditActivity isn't the view, and that's kind of the whole point.  an Activity is a collection of 'stacks' of M/V/VM components

exallium commented on app/src/main/java/com/exallium/mvvmcleanexample/app/binding/TextInputLayoutBindingAdapters.kt:
```diff
@@ -0,0 +1,12 @@
+private const val BIND_ERROR = "bind:error"
+
+@BindingAdapter(BIND_ERROR)
+fun TextInputLayout.bindTextInputLayoutError(error: String?) {
```
Gettin' fancy.  Let's add a note here about why this works.

Aidanvii7 replied:
Good point, will do!

exallium commented on app/src/main/java/com/exallium/mvvmcleanexample/app/users/UserEditActivity.kt:
```diff
@@ -14,45 +14,57 @@ import com.exallium.mvvmcleanexample.app.users.di.DaggerUserEditComponent
-                is AppRouter.Transition.GoBack -> onBackPressed()
-                is AppRouter.Transition.DisplayMessage -> displayMessage(it.message)
-                is AppRouter.Transition.DisplayError -> displayError(binding.root, it.message)
+            router.transitions().subscribe {
```
if we're adding this extra step of scope, let's refactor some of the stuff in this method out into their own methods (like an inject method and a setupRouter method)

exallium commented on presentation/build.gradle:
```diff
@@ -37,6 +38,7 @@ dependencies {
     compile project(path: ':domain')
     compile "com.jakewharton.rxrelay2:rxrelay:2.0.0"
     testCompile "com.nhaarman:mockito-kotlin:$mockito_version"
+    kapt "com.android.databinding:compiler:$android_gradle_version"
```
was this required here? There shouldn't be any BindingAdapters in the presentation layer, only fields.  There isn't anything that should have to be generated.

Aidanvii7 replied:
You're right​, will remove and move the adapters to the app layer.

exallium commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/users/UserEditViewModel.kt:
```diff
@@ -1,61 +1,93 @@
-        compositeDisposable.add(obs.subscribe(firstNameRelay))
+    // view >> view model
+    fun saveUser() {
+        savedUser = User(firstName, lastName)
```
I'm not in love with `savedUser`.  It isn't clear here via naming that we're performing an action

exallium commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/utils/Extensions.kt:
```diff
@@ -0,0 +1,178 @@
+ * @param onChangeOnInit if true, the provided [onChange] block will be executed with the [initialValue]. The default is true.
+ *
+ */
+fun <T> distinctObservable(
```
this is looking better

exallium reviewed (changes requested):
Couple of things left.  Starting to look good =) Let's make sure we've got things in the right modules (just double check).

Ideally, the app module is the only one that ever imports View or anything in android not related to databinding, if that makes sense.

Aidanvii7 commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/utils/delegates/BindableProperty.kt:
```diff
@@ -0,0 +1,33 @@
+
+    abstract fun notify(bindable: B, propertyId: Int, newValue: T)
+
+    companion object{
```
TODO move to top

exallium commented on app/src/main/java/com/exallium/mvvmcleanexample/app/users/UserEditActivity.kt:
```diff
@@ -40,19 +44,31 @@ class UserEditView : AppCompatActivity() {
+
+    private fun onNextTransition(transition: AppRouter.Transition) {
+        when (transition) {
+            is AppRouter.Transition.GoBack -> finish()
```
`finis()` ... of course!

Aidanvii7 replied:
I dont think you are supposed to invoke `onBackPressed()` programatically, this was causing a crash for me.

exallium replied:
you're totally right.

exallium commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/utils/delegates/EcoCompositeDisposableProperty.kt:
```diff
@@ -0,0 +1,22 @@
+ * When a new value is assigned, [setValue] will call [Disposable.dispose] on the previous value if available.
+ * @param disposable the new value
+ */
+class EcoCompositeDisposableProperty(private var disposable: CompositeDisposable) : ReadWriteProperty<Any?, CompositeDisposable> {
```
why "Eco" ?

Aidanvii7 replied:
Maybe not the best name, couldn't think of anything else short that made sense.
These 'eco' property delegates are intended to perform some kind of cleanup of their previous value on reassignment. 

Though I think the 'eco' prefix can be dropped come to think of it.

exallium commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/utils/delegates/EcoDisposableProperty.kt:
```diff
@@ -0,0 +1,21 @@
+ * When a new value is assigned, [setValue] will call [Disposable.dispose] on the previous value if available.
+ * @param disposable the new value
+ */
+class EcoDisposableProperty(private var disposable: Disposable?) : ReadWriteProperty<Any?, Disposable?> {
```
why "Eco"?

exallium reviewed (changes requested):
This is really close.  I've got a few questions I'd like answered, but other than that I think this is good.  I want to make sure the codebase is as clean and understandable as possible, you shouldn't need to be an expert in Kotlin to get the gist of what's going on.

Aidanvii7 commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/users/UserEditViewModel.kt:
```diff
@@ -1,61 +1,92 @@
+    // view << view model
+    @get:Bindable
+    var firstNameError: String by bindable("")
+        private set
```
The property can now only be changed internally, this is not possible with `ObservableField` etc.

exallium replied:
neat.

Aidanvii7 commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/utils/Extensions.kt:
```diff
@@ -0,0 +1,241 @@
+inline fun <T> BaseObservable.bindable(
+        initialValue: T,
+        bindableResourceId: Int = USE_PROPERTY_CACHER,
+        crossinline onChange: OnChange<T>,
```
Ideally this function should be the only implementation (where the one above disappears).

Due to a limitation with current Kotlin, we cannot provide defaults for inlined lambda expressions, the following yields an error:
`onChange: OnChange<T> = {}  // construction is not yet supported in inline functions`

Aidanvii7 commented on presentation/src/main/java/com/exallium/mvvmcleanexample/presentation/users/UserEditViewModel.kt:
```diff
@@ -1,61 +1,92 @@
-    private val lastNameRelay = BehaviorRelay.create<String>()
+    // view |< >> view model
+    var firstName: String by distinctObservable(
+            initialValue = "",
```
The view will take the initial value from the `firstName` property, but if the view model changes this property it will not be propagated to the view. Once the initial value has been read by the view (view binding passes it to the view I mean), the view (binding) will simply write to the property, and the view model will observe the changes.

Aidanvii7 commented:
@exallium I understand, there is some advanced Kotlin going on here. If you're not comfortable with the property delegates, I could remove them. It just means there's going to be a whole lot more code in their place. If you want I could put comments in the places where they are used, on top of the documentation.

exallium commented:
@Aidanvii7 As well as it's well documented, concise, and easy to follow, I'm fine with it.  Let's make sure it's very clear how things work and why we chose certain approaches.

The advanced features are great, and I'm familiar and comfortable with them, but the same might not be true for a more junior developer looking to get their feet wet.  So, let's document where we can. (Let's not remove a bunch of code if it's just going to result in adding more.)

Aidanvii7 committed:
Reworked view model and BindingAdapters, added property delegates

Added property delegates to handle some boilerplate code, with
extension functions to make use of them.

Removed rx from the view<>view model bindings - rx is now only used
behind the view model, like standard BindingAdapters from the framework.

Text BindingAdapter is now a simplified version of the framework
equivalent, optionally supporting inverse bindings.

BindingAdapters are now extension functions for their respective View
type.

Also added project settings for tests to ignore Kotlin identifiers that are considered illegal in Android.

exallium commented:
Let's get this in, and we can discuss and iterate if we want more changes made.
